keps/sig-cluster-lifecycle/kubeadm/20190506-kubeadm-machine-output.md

```diff
@@ -0,0 +1,370 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Kubeadm should support structured output, such as JSON, YAML, or a [Go template](https://golang.org/pkg/text/template/).
+
+## Motivation
+
+While Kubernetes may be deployed manually, the de facto, if not de jeur, means of turning up a Kubernetes cluster is kubeadm. Popular systems management software, such as Terraform, rely on kubeadm in order to deploy Kubernetes. Planned enhancements to the Cluster API project include a composable package for bootstrapping Kubernetes with kubeadm and cloud-init.
+
+Without structured output, even the most seemingly innocuous changes could break Terraform, Cluster API, or other software that relies on the results of kubeadm.
+
+### Goals
+
+* Kubeadm should support structured output, including, but not limited to, the following output types:
+  * Buffered
+    * JSON
+    * YAML
+    * Go template
+  * Unbuffered
+    * Text (the current behavior)
+  * Please see [*Buffered vs Unbuffered*](#buffered-vs-unbuffered) for details on what both terms mean in the context of this KEP
+* Kubeadm should support the above output types for commands that include, but not necessarily limited to, the following:
+  * `alpha certs`
+  * `config images list`
+  * `init`
+  * `token create`
+  * `token list`
+  * `upgrade plan`
+  * `version`
+
+### Non-Goals
+
+* Explicit support for an unbuffered, *short* variant of *Text* output is not necessary at this time as the only command that currently supports this output type is `version`
+* [Parity with kubectl's `-o|--output` flag](#parity-with-kubectl-and-versioned-output)
+* [Versioned output](#parity-with-kubectl-and-versioned-output)
+* [Using the printers](#parity-with-kubectl-and-versioned-output) from the `kubernetes/cli-runtime` package
+
+## Proposal
+
+### User Stories
+
+The examples in the user stories are predicated on:
+
+* [Awareness of `jq`](#jq)
+* [A friendly bootstrap token struct](#a-friendly-bootstrap-token-struct)
+* [Additional functions for Go templates](#go-template-functions)
+
+#### Story 1
+
+The deployment of a multi-node Kubernetes cluster is automated by parsing [the JSON output](#kubeadm-init-json-output) of `kubeadm init` for the emitted `kubeadm join` command.
+
+1. Run `kubeadm init -o json | jq -r '"kubeadm join \(.node0) --token \(.token.id) --discovery-token-ca-cert-hash \(.caCrt)"'`
+
+#### Story 2
+
+A script requires a list of token IDs.
+
+1. Run `kubeadm token list -o '{{range .}}{{println .ID}}{{end}}'`
```

Comment:
I think to avoid ambiguity and to allow for other template syntaxes (jsonpath & go templates) kubectl uses `-o go-template={{range...`